lib/internal/Magento/Framework/App/Cache/FlushCacheByTags.php

```diff
@@ -56,36 +50,32 @@ public function __construct(
     }
 
     /**
-     * Clean cache on save object
+     * Clean cache when object is saved
      *
-     * @param AbstractResource $subject
-     * @param \Closure $proceed
-     * @param AbstractModel $object
-     * @return AbstractResource
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function aroundSave(AbstractResource $subject, \Closure $proceed, AbstractModel $object): AbstractResource
-    {
-        $result = $proceed($object);
+    public function afterSave(
+        AbstractResource $subject,
+        AbstractResource $result,
+        AbstractModel $object
+    ): AbstractResource {
         $tags = $this->tagResolver->getTags($object);
         $this->cleanCacheByTags($tags);
 
         return $result;
     }
 
     /**
-     * Clean cache on delete object
+     * Clean cache when object is deleted
      *
-     * @param AbstractResource $subject
-     * @param \Closure $proceed
-     * @param AbstractModel $object
-     * @return AbstractResource
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function aroundDelete(AbstractResource $subject, \Closure $proceed, AbstractModel $object): AbstractResource
-    {
+    public function afterDelete(
+        AbstractResource $subject,
+        AbstractResource $result,
+        AbstractModel $object
+    ): AbstractResource {
         $tags = $this->tagResolver->getTags($object);
-        $result = $proceed($object);
```

Comment:
The tags are retrieved before deleting the Resource, not after. But we need to Purge the cache only after the delete. If the delete fails, it should rollback the transaction and not commit, hence not purge the cache. This needs to stay an around.

Reply:
A rollback will throw an Exception, right?  In which case the line after $proceed is not called.  This same scenario occurs with the after; an exception is thrown and the after is not called.

I failed to find any instance in which the tags changed after an object was deleted.  This would be the only reason to call getTags before deletion.

Reply:
I'm not sure if the entityid to crate the tag will still be available on the object once it's deleted. Need to debug and test that to make sure the tag will be created the same way once the object is deleted. The object might be a product, saleable item, category, cms page, etc... This needs to account for all these types.